src/engine/LocalVocab.h

```diff
@@ -0,0 +1,69 @@
+// Copyright 2022, University of Freiburg
+// Chair of Algorithms and Data Structures
+// Author: Hannah Bast <bast@cs.uni-freiburg.de>
+
+#pragma once
+
+#include "absl/container/node_hash_map.h"
+
+// A class for maintaing a local vocabulary with contiguous (local) IDs. This is
+// meant for words that are not part of the normal vocabulary (constructed from
+// the input data at indexing time).
+//
+// TODO: This is a first version of this class with basic functionality. Note
+// that the local vocabulary used to be a simple `std::vector<std::string>`
+// defined inside of the `ResultTable` class. You gotta start somewhere.
+class LocalVocab {
+ public:
+  // Create a new, empty local vocabulary.
+  LocalVocab() = default;
+
+  // Prevent accidental copying of a local vocabulary (it can be quite large),
+  // but moving it is OK.
+  //
+  // TODO: does the default move do the "right" thing, that is, move the hash
+  // map instead of copying it?
+  LocalVocab(const LocalVocab&) = delete;
+  LocalVocab(LocalVocab&&) = default;
+
+  // Get ID of a word in the local vocabulary. If the word was already
+  // contained, return the already existing ID. If the word was not yet
+  // contained, add it, and return the new ID.
+  Id getIdAndAddIfNotContained(const std::string& word);
```

Comment:
Or maybe even better
`string_view` and `string&&`.
This gives us the most flexibility.
The only disadvantage is, that passing in `const char*` doesn't work, because it is ambiguous. We could either

- explicitly cast all the `const char*` to `string_view` when passing them in.
- Add a third overload that takes `const char*`

[Godbolt Link for playing around](https://godbolt.org/z/zdWdWPq1z)

Reply:
Edit:
`string_view` is not that easy,
as the hashMap only works with and hashes strings, so just go with `const string&` and `string&&` as initially suggested.